### PR TITLE
Fix UGENE-5371

### PR DESCRIPTION
**Pull request by ichebyki**
Fix UGENE-5371
Add proxy functions for open/fopen functions in samtools library
The functions accept multibyte char* string convert it to wide char* string and call _wopen/_wfopen

### CODE REVIEW

File: src/corelibs/U2Formats/src/BAMUtils.cpp [hunk 1]
```diff
@@ -155,9 +155,13 @@ namespace {
 void BAMUtils::convertToSamOrBam(const GUrl &samUrl, const GUrl &bamUrl, const ConvertOption &options, U2OpStatus &os ) {
     const QByteArray samFileName = samUrl.getURLString().toLocal8Bit();
     const QByteArray bamFileName = bamUrl.getURLString().toLocal8Bit();
+    const char* sam_path = samUrl.getURLStringAnsi();
+    const char* bam_path = bamUrl.getURLStringAnsi();
```
Comment (DmitriiSukhomlinov) [inline src/corelibs/U2Formats/src/BAMUtils.cpp]: We use lowerCamelCase for variable naming, not snake_case

File: src/corelibs/U2Formats/src/BAMUtils.cpp [hunk 2]
```diff
@@ -155,9 +155,13 @@ namespace {
...
 
     QByteArray sourceName = (options.samToBam)? samFileName : bamFileName;
     QByteArray targetName = (options.samToBam)? bamFileName : samFileName;
+    const char* source_path = (options.samToBam) ? sam_path : bam_path;
+    const char* target_path = (options.samToBam) ? bam_path : sam_path;
```
Comment (DmitriiSukhomlinov) [inline src/corelibs/U2Formats/src/BAMUtils.cpp]: The same

File: src/corelibs/U2Formats/src/BAMUtils.cpp [hunk 3]
```diff
@@ -601,6 +603,7 @@ void BAMUtils::writeObjects(const QList<GObject*> &objects, const GUrl &urlStr,
 
     QByteArray url = urlStr.getURLString().toLocal8Bit();
     CHECK_EXT(!url.isEmpty(), os.setError("Empty file url"), );
+    const char* url_path = urlStr.getURLStringAnsi();
```
Comment (DmitriiSukhomlinov) [inline src/corelibs/U2Formats/src/BAMUtils.cpp]: The same

File: src/libs_3rdparty/samtools/src/ugene_custom_io.c [hunk 4]
```diff
@@ -0,0 +1,63 @@
+#include <io.h>
+#include <stdio.h>
+#include <windows.h>
```
Comment (DmitriiSukhomlinov) [inline src/libs_3rdparty/samtools/src/ugene_custom_io.c:4]: If the OS is not Windows will it work?

File: src/libs_3rdparty/zlib/src/ugene_custom_io.c [hunk 5]
```diff
@@ -0,0 +1,63 @@
+#include <io.h>
+#include <stdio.h>
+#include <windows.h>
```
Comment (DmitriiSukhomlinov) [inline src/libs_3rdparty/zlib/src/ugene_custom_io.c:4]: And here?

**Review by DmitriiSukhomlinov — CHANGES_REQUESTED** [review]
First of all, as far as I understood, you've changed the behavior for Windows only, and what about Linux and Mac? 
Also, please, check naming, we use lowerCamelCase almost for everything.

File: src/corelibs/U2Core/src/globals/GUrl.cpp [hunk 6]
```diff
@@ -143,6 +147,26 @@ bool GUrl::operator !=(const GUrl& url) const {
...
+// The function converts url string to multibyte form
+// default code page is CP_THREAD_ACP
+const char* GUrl::getURLStringAnsi(int codePage) const {
+#ifdef Q_OS_WIN
+    std::wstring wPath = getURLString().toStdWString();
+    codePage = codePage < 0 ? CP_THREAD_ACP : codePage;
+
+    DWORD buffSize = WideCharToMultiByte(codePage, 0, wPath.c_str(), -1, NULL, 0, NULL, NULL);
+    if (!buffSize)
+        return NULL;
```
Comment (atiunov) [inline src/corelibs/U2Core/src/globals/GUrl.cpp]: Our standard is c++11 now. Try to use nullptr instead of NULL. It is not critical, of course.

**Review by atiunov — APPROVED** [review]

File: src/corelibs/U2Core/src/globals/GUrl.cpp (hunk 6 above)
Comment (ichebyki) [reply src/corelibs/U2Core/src/globals/GUrl.cpp]: done

**Pull request merged**